File: Jasmine/JasmineCompiler/Src/DFIR_Compiler/DFIR_Compiler/IPrimitive.cs
using System;
using System.Collections.Generic;


namespace DFIR_Compiler
{
    public class IPrimitive:INode
    {
        private IPrimitiveMode mode;
        public IPrimitive(int nodeId, int parentId, List<ITerminal> inputTerminals, List<ITerminal> outputTerminals, IPrimitiveMode mode) : base(nodeId, parentId, inputTerminals, outputTerminals)
        {
            this.mode = mode;
            INodeType = "IPrimitive";
        }

        public virtual IPrimitiveMode GetMode()
        {
            return mode;
        }

        public virtual string GetNameByOutputIndex(int index)
        {
            string res = "";
            res = GetName() + '_' + Convert.ToString(index);
            return res;
        }
        public virtual string GetNameByOutputConnection(INode n)
        {
            string res = "";
            res = GetNameByOutputIndex(GetOutputConnectionIndex(n));
            return res;
        }

        //The IPrimitive Function
        public virtual List<string> ExSelectPrimitive()
        {
            List<string> res = new List<string>();
            res.Add(Indentation(1) + "if(x1 == true)");
            res.Add(Indentation(2) + "y0 = x2;");
            res.Add(Indentation(1) + "else");
            res.Add(Indentation(2) + " y0 = x0;");
            return res;
        }
        public virtual List<string> ExIncrementPrimitive()
        {
            List<string> res = new List<string>();
            res.Add(Indentation(1) + "y0 = x0+1;");
            return res;
        }
        public virtual List<string> ExDecrementPrimitive()
        {
            List<string> res = new List<string>();
            res.Add(Indentation(1) + "y0 = x0-1;");
            return res;
        }
        public virtual List<string> ExAbsoluteValuePrimitive()
        {
            List<string> res = new List<string>();
            IDiagram.setHasCmath(true);
            res.Add(Indentation(1) + "y0 = abs(x0);");
            return res;
        } 
        public virtual List<string> ExAddArrayElementsPrimitive()
        {
            List<string> res = new List<string>();
            res.Add(Indentation(1) + "y0 = x0-1;");
            return res;
        }
        public virtual List<string> ExScaleByPowerOf2Primitive()
        {
            List<string> res = new List<string>();
            IDiagram.setHasCmath(true);
            res.Add(Indentation(1) + "y0 = x0 * pow(2.0, x1);");
            return res;
        }
        public virtual List<string> ExSquareRootPrimitive()
        {
            List<string> res = new List<string>();
            IDiagram.setHasCmath(true);
            res.Add(Indentation(1) + "y0 = sqrt(x0);");
            return res;
        }
        public virtual List<string> ExSquarePrimitive()
        {
            List<string> res = new List<string>();
            res.Add(Indentation(1) + "y0 = x0*x0;");
            return res;
        }
        public virtual List<string> ExNegatePrimitive()
        {
            List<string> res = new List<string>();
            res.Add(Indentation(1) + "y0 = -x0;");
            return res;
        }
        public virtual List<string> ExReciprocalPrimitive()
        {
            List<string> res = new List<string>();
            res.Add(Indentation(1) + "y0 = 1/x0;");
            return res;
        }
        public virtual List<string> ExMaxAndMinPrimitive()
        {
            List<string> res = new List<string>();
            res.Add(Indentation(1) + "if(x0 >= x1){");
            res.Add(Indentation(2) + "y0 = x1;");
            res.Add(Indentation(2) + "y1 = x0;");
            res.Add(Indentation(1) + "}else{");
            res.Add(Indentation(2) + "y0 = x0;");
            res.Add(Indentation(2) + "y1 = x1;");
            res.Add(Indentation(1) + "}");
            return res;
        }
        public virtual List<string> ExDecrementPrimitive1()
        {
            List<string> res = new List<string>();
            res.Add(Indentation(1) + "return p0-1;");
            return res;
        }

        public virtual string Indentation(int indnt)
        {
            string s = "";
            for (int i = 0; i < indnt; i++)
            {
                s += "  ";
            }
            return s;
        }

        public virtual void GetIPrimitiveCCode(List<string> res)
        {
            if (mode == IPrimitiveMode.ExSelectPrimitive)
            {
                res.AddRange(ExSelectPrimitive());
            }
            else if (mode == IPrimitiveMode.ExIncrementPrimitive)
            {
                res.AddRange(ExIncrementPrimitive());
            }
            else if (mode == IPrimitiveMode.ExDecrementPrimitive)
            {
                res.AddRange(ExDecrementPrimitive());
            }
            else if (mode == IPrimitiveMode.ExAbsoluteValuePrimitive)
            {
                res.AddRange(ExAbsoluteValuePrimitive());
            }
            else if (mode == IPrimitiveMode.ExAddArrayElementsPrimitive)
            {
                res.AddRange(ExAddArrayElementsPrimitive());
            }
            else if (mode == IPrimitiveMode.ExScaleByPowerOf2Primitive)
            {
                res.AddRange(ExScaleByPowerOf2Primitive());
            }
            else if (mode == IPrimitiveMode.ExSquareRootPrimitive)
            {
                res.AddRange(ExSquareRootPrimitive());
            }
            else if (mode == IPrimitiveMode.ExSquarePrimitive)
            {
                res.AddRange(ExSquarePrimitive());
            }
            else if (mode == IPrimitiveMode.ExNegatePrimitive)
            {
                res.AddRange(ExNegatePrimitive());
            }
            else if (mode == IPrimitiveMode.ExReciprocalPrimitive)
            {
                res.AddRange(ExReciprocalPrimitive());
            }
            else if (mode == IPrimitiveMode.ExMaxAndMinPrimitive)
            {
                res.AddRange(ExMaxAndMinPrimitive());
            }
        }

        public override void Print()
        {
            Console.WriteLine("------------------------------------");
            Console.WriteLine("IPrimitive:");
            Console.WriteLine("NodeId: " + nodeId + ' ' + "ParentId: " + parentId + "\nMode: " + mode);
            for (int i = 0; i < inputTerminals.Count; i++)
            {
                inputTerminals[i].Print();
            }
            for (int i = 0; i < outputTerminals.Count; i++)
            {
                outputTerminals[i].Print();
            }
        }

    }
}

File: Wayne/WayneCompiler/src/Dfir_representation/IPrimitive.java
package Dfir_representation;

import java.util.Vector;


public class IPrimitive extends JNode{
	   
	   private IPrimitiveMode mode;
	   public IPrimitive(int nodeId, int parentId, Vector<ITerminal> inputTerminals,Vector<ITerminal> outputTerminals, IPrimitiveMode mode){
		   super(nodeId,parentId,inputTerminals,outputTerminals);
		   this.mode = mode;
		   JNodeType = "IPrimitive";
	   }
	   public IPrimitiveMode getMode(){
		   return this.mode;
	   }
	   public void getCCode(Vector <String> res){
		   if(mode == IPrimitiveMode.valueOf("ExSelectPrimitive")){
				res.addAll(ExSelectPrimitive());
			}else if(mode == IPrimitiveMode.valueOf("ExIncrementPrimitive")){
				res.addAll(ExIncrementPrimitive());
			}else if(mode == IPrimitiveMode.valueOf("ExDecrementPrimitive")){
				res.addAll(ExDecrementPrimitive());
			}else if(mode == IPrimitiveMode.valueOf("ExAbsoluteValuePrimitive")){
				res.addAll(ExAbsoluteValuePrimitive());
			}
			else if(mode == IPrimitiveMode.valueOf("ExAddArrayElementsPrimitive")){
				res.addAll(ExAddArrayElementsPrimitive());
			}
			else if(mode == IPrimitiveMode.valueOf("ExScaleByPowerOf2Primitive")){
				res.addAll(ExScaleByPowerOf2Primitive());
			}
			else if(mode == IPrimitiveMode.valueOf("ExSquareRootPrimitive")){
				res.addAll(ExSquareRootPrimitive());
			}
			else if(mode == IPrimitiveMode.valueOf("ExSquarePrimitive")){
				res.addAll(ExSquarePrimitive());
			}
			else if(mode == IPrimitiveMode.valueOf("ExNegatePrimitive")){
				res.addAll(ExNegatePrimitive());
			}
			else if(mode == IPrimitiveMode.valueOf("ExReciprocalPrimitive")){
				res.addAll(ExReciprocalPrimitive());
			}
			else if(mode == IPrimitiveMode.valueOf("ExMaxAndMinPrimitive")){
				res.addAll(ExMaxAndMinPrimitive());
			}
	   }
	   public String getNameByOutputIndex(int index) {
			String res = "";
			res = this.getName()+'_'+Integer.toString(index);
			return res;
		}
	   public String getNameByOutputConnection(JNode n) {
			String res = "";
			res = this.getNameByOutputIndex(this.getOutputConnectionIndex(n));
			return res;
		}
	   public Vector<String> ExSelectPrimitive() { 
			Vector<String> res = new Vector<String>();
			res.addElement(indent(1) + "if(p1 == true) q0 = p2;");
			res.addElement(indent(1) + "else q0 = p0;");
			return res;
		}
		public Vector<String> ExIncrementPrimitive() { 
			Vector<String> res = new Vector<String>();
			res.addElement(indent(1) + "q0 = p0+1;");
			return res;
		}
		public Vector<String> ExDecrementPrimitive() { 
			Vector<String> res = new Vector<String>();
			res.addElement(indent(1) + "q0 = p0-1;");
			return res;
		}
		public Vector<String> ExAbsoluteValuePrimitive() { 
			Vector<String> res = new Vector<String>();
			IDiagram.setHasCmath(true);
			res.addElement(indent(1) + "q0 = abs(p0);");
			return res;
		}public Vector<String> ExAddArrayElementsPrimitive() { // problem with the array size
			Vector<String> res = new Vector<String>();
			res.addElement(indent(1) + "q0 = p0-1;");
			return res;
		}public Vector<String> ExScaleByPowerOf2Primitive() { 
			Vector<String> res = new Vector<String>();
			IDiagram.setHasCmath(true);
			res.addElement(indent(1) + "q0 = p0 * pow(2.0, p1);");
			return res;
		}public Vector<String> ExSquareRootPrimitive() { 
			Vector<String> res = new Vector<String>();
			IDiagram.setHasCmath(true);
			res.addElement(indent(1) + "q0 = sqrt(p0);");
			return res;
		}public Vector<String> ExSquarePrimitive() { 
			Vector<String> res = new Vector<String>();
			res.addElement(indent(1) + "q0 = p0*p0;");
			return res;
		}public Vector<String> ExNegatePrimitive() { 
			Vector<String> res = new Vector<String>();
			res.addElement(indent(1) + "q0 = -p0;");
			return res;
		}public Vector<String> ExReciprocalPrimitive() { 
			Vector<String> res = new Vector<String>();
			res.addElement(indent(1) + "q0 = 1/p0;");
			return res;
		}public Vector<String> ExMaxAndMinPrimitive() { 
			Vector<String> res = new Vector<String>();
			res.addElement(indent(1) + "if(p0 >= p1){");
			res.addElement(indent(2) + "q0 = p1;");
			res.addElement(indent(2) + "q1 = p0;");
			res.addElement(indent(1) + "}else{");
			res.addElement(indent(2) + "q0 = p0;");
			res.addElement(indent(2) + "q1 = p1;");
			res.addElement(indent(1) + "}");
			return res;
		}public Vector<String> ExDecrementPrimitive1() { 
			Vector<String> res = new Vector<String>();
			res.addElement(indent(1) + "return p0-1;");
			return res;
		}
		
		public String indent(int indnt){
			String s = "";
			for(int i=0; i<indnt; i++){
				s += "  ";
			}
			return s;
		}
		
	   public void print(){
		   System.out.println("IPrimitive begin");
		   System.out.println("NodeId: " + this.nodeId + ' ' + "ParentId: " + this.parentId + "\nMode: " + this.mode);
		   for(int i=0; i<inputTerminals.size(); i++){
				   inputTerminals.get(i).print();
		   }
		  for(int i=0; i<outputTerminals.size(); i++){
				   outputTerminals.get(i).print();
		   }   
		  System.out.println("IPrimitive end\n");
	   }
}
